src/game/Tools/Formulas.h

```diff
@@ -297,61 +297,62 @@ namespace MaNGOS
             return GRAY;
         }
 
-        inline uint32 GetZeroDifference(uint32 pl_level)
+        inline uint32 GetZeroDifference(uint32 unit_level)
         {
-            if (pl_level < 8)  return 5;
-            if (pl_level < 10) return 6;
-            if (pl_level < 12) return 7;
-            if (pl_level < 16) return 8;
-            if (pl_level < 20) return 9;
-            if (pl_level < 30) return 11;
-            if (pl_level < 40) return 12;
-            if (pl_level < 45) return 13;
-            if (pl_level < 50) return 14;
-            if (pl_level < 55) return 15;
-            if (pl_level < 60) return 16;
+            if (unit_level < 8)  return 5;
+            if (unit_level < 10) return 6;
+            if (unit_level < 12) return 7;
+            if (unit_level < 16) return 8;
+            if (unit_level < 20) return 9;
+            if (unit_level < 30) return 11;
+            if (unit_level < 40) return 12;
+            if (unit_level < 45) return 13;
+            if (unit_level < 50) return 14;
+            if (unit_level < 55) return 15;
+            if (unit_level < 60) return 16;
             return 17;
         }
 
-        inline uint32 BaseGain(uint32 pl_level, uint32 mob_level)
+        inline float BaseGain(uint32 unit_level, uint32 mob_level)
         {
-            const uint32 nBaseExp = 45;
-            if (mob_level >= pl_level)
+            const uint32 nBaseExp = unit_level * 5 + 45;
+            if (mob_level >= unit_level)
             {
-                uint32 nLevelDiff = mob_level - pl_level;
+                uint32 nLevelDiff = mob_level - unit_level;
                 if (nLevelDiff > 4)
                     nLevelDiff = 4;
-                return ((pl_level * 5 + nBaseExp) * (20 + nLevelDiff) / 10 + 1) / 2;
+                return nBaseExp * (1.0f + (0.05f * nLevelDiff));
             }
-            uint32 gray_level = GetGrayLevel(pl_level);
+            uint32 gray_level = GetGrayLevel(unit_level);
             if (mob_level > gray_level)
             {
-                uint32 ZD = GetZeroDifference(pl_level);
-                return (pl_level * 5 + nBaseExp) * (ZD + mob_level - pl_level) / ZD;
+                uint32 ZD = GetZeroDifference(unit_level);
+                uint32 nLevelDiff = unit_level - mob_level;
+                return nBaseExp * (1.0f - (float(nLevelDiff) / ZD));
             }
             return 0;
         }
 
-        inline uint32 Gain(Player* player, Creature* target)
+        inline uint32 Gain(Unit* unit, Creature* target)
```

Comment:
Masybe possible use Unit const* for avoid static casts?

Reply:
Not quite sure what you're saying, so if I change this inline to use Unit const* instead of Unit* then I don't have to do static casting from the other functions calling this? so I can just pass whatever instead (pet/player)? Wouldn't I still have to cast them as unit?

Reply:
Yes, if in function 'unit' not use non-const functions

Reply:
Владимир, можешь по-русски, а я переведу? :>

Reply:
Why need any explicit cast code in this function call? Just pass Pet* to arg.

Reply:
Really?

Reply:
I see you drop explict casts. Now this is more clean code.